File: MPC模型预测控制器-4-Octave完整案例讲解/code/python/mpc_matrices.py
"""构造无约束线性 MPC 的预测矩阵和二次型代价矩阵。"""

from __future__ import annotations

import numpy as np
from numpy.typing import NDArray


FloatArray = NDArray[np.float64]


def mpc_matrices(
    a: FloatArray,
    b: FloatArray,
    q: FloatArray,
    r: FloatArray,
    f: FloatArray,
    horizon: int,
) -> tuple[FloatArray, FloatArray]:
    """返回代价函数所需的 E、H 矩阵。

    预测模型写成 ``X = M @ x(k) + C @ U``，与控制序列有关的
    代价为 ``U.T @ H @ U + 2 * x(k).T @ E.T @ U``。
    """
    a = np.asarray(a, dtype=float)
    b = np.asarray(b, dtype=float)
    q = np.asarray(q, dtype=float)
    r = np.asarray(r, dtype=float)
    f = np.asarray(f, dtype=float)

    if a.ndim != 2 or a.shape[0] != a.shape[1]:
        raise ValueError("A 必须是方阵")

    n = a.shape[0]
    if b.ndim != 2 or b.shape[0] != n:
        raise ValueError("A 与 B 的行数必须一致")

    p = b.shape[1]
    if q.shape != (n, n):
        raise ValueError("Q 的尺寸必须是 n x n")
    if f.shape != (n, n):
        raise ValueError("F 的尺寸必须是 n x n")
    if r.shape != (p, p):
        raise ValueError("R 的尺寸必须是 p x p")
    if not isinstance(horizon, int) or horizon < 1:
        raise ValueError("预测区间 horizon 必须是正整数")

    # 当前状态到 x(k), ..., x(k+N) 的映射。
    m = np.zeros(((horizon + 1) * n, n))
    m[:n, :] = np.eye(n)

    # 未来控制序列到预测状态序列的映射。
    c = np.zeros(((horizon + 1) * n, horizon * p))
    a_power = np.eye(n)

    for i in range(1, horizon + 1):
        rows = slice(i * n, (i + 1) * n)
        previous_rows = slice((i - 1) * n, i * n)

        c[rows, :p] = a_power @ b
        if horizon > 1:
            c[rows, p:] = c[previous_rows, :-p]

        a_power = a @ a_power
        m[rows, :] = a_power

    q_bar = np.zeros(((horizon + 1) * n, (horizon + 1) * n))
    for i in range(horizon):
        rows = slice(i * n, (i + 1) * n)
        q_bar[rows, rows] = q
    q_bar[horizon * n :, horizon * n :] = f

    r_bar = np.kron(np.eye(horizon), r)

    e = c.T @ q_bar @ m
    h = c.T @ q_bar @ c + r_bar

    # 消除浮点运算产生的微小非对称。
    h = (h + h.T) / 2.0
    return e, h

File: deprecated/MPC模型预测控制器-4-Octave完整案例讲解/code/octave/MPC_Matrices.m
function [E, H] = MPC_Matrices(A, B, Q, R, F, N)
%MPC_MATRICES 构造无约束线性 MPC 的二次规划矩阵。
%
% 预测模型：
%   X = M*x(k) + C*U
%
% 代价函数中与 U 有关的部分：
%   U'*H*U + 2*x(k)'*E'*U

    n = size(A, 1);
    p = size(B, 2);

    assert(size(A, 2) == n, 'A 必须是方阵。');
    assert(size(B, 1) == n, 'A 与 B 的行数必须一致。');
    assert(isequal(size(Q), [n, n]), 'Q 的尺寸必须是 n x n。');
    assert(isequal(size(F), [n, n]), 'F 的尺寸必须是 n x n。');
    assert(isequal(size(R), [p, p]), 'R 的尺寸必须是 p x p。');
    assert(isscalar(N) && N >= 1 && N == fix(N), ...
           'N 必须是正整数。');

    % M 将当前状态映射到 x(k),...,x(k+N)。
    M = zeros((N + 1) * n, n);
    M(1:n, :) = eye(n);

    % C 将未来控制序列映射到预测状态序列。
    C = zeros((N + 1) * n, N * p);

    A_power = eye(n);
    for i = 1:N
        rows = i * n + (1:n);
        previous_rows = (i - 1) * n + (1:n);

        C(rows, :) = [A_power * B, C(previous_rows, 1:end-p)];
        A_power = A * A_power;
        M(rows, :) = A_power;
    end

    Q_bar = blkdiag(kron(eye(N), Q), F);
    R_bar = kron(eye(N), R);

    E = C' * Q_bar * M;
    H = C' * Q_bar * C + R_bar;

    % 消除浮点运算造成的微小非对称，避免部分求解器拒绝 H。
    H = (H + H') / 2;
end
